src/io/GammaWriter.cpp

```diff
@@ -5,103 +5,119 @@
+    for (unsigned componentId = 0; componentId < _numComp; ++componentId) {
+        _gamma.at(componentId) = 0.0;
+    }
 }
 
-void GammaWriter::resetGamma() {
-	for(unsigned componentId = 0; componentId < _simulation.getEnsemble()->getComponents()->size(); ++componentId) {
-		_Gamma[componentId] = 0;
-	}
+inline double GammaWriter::getGamma(unsigned id) {
+    return (_gamma.at(id)/(_globalLength.at(0)*_globalLength.at(2)*_numInterfaces));
 }
 
-double GammaWriter::getGamma(unsigned id, double globalLength[3]){
-	return (_Gamma[id]/(2*globalLength[0]*globalLength[2]));
-}
-
-void GammaWriter::calculateGamma(ParticleContainer* particleContainer, DomainDecompBase* domainDecom){
-	unsigned numComp = _simulation.getEnsemble()->getComponents()->size();
-	std::vector<double> _localGamma(numComp, 0.0);
-
-	#if defined(_OPENMP)
-	#pragma omp parallel
-	#endif
-	{
-
-		std::vector<double> localGamma_thread(numComp, 0.0);
-
-		for (auto tempMol = particleContainer->iterator(ParticleIterator::ONLY_INNER_AND_BOUNDARY); tempMol.isValid(); ++tempMol) {
-
-			if (particleContainer->isInBoundingBox(tempMol->r_arr().data())) {
-				unsigned cid = tempMol->componentid();
-				localGamma_thread[cid] += tempMol->Vi(1) - 0.5 * (tempMol->Vi(0) + tempMol->Vi(2));
-			}
-		}
-
-		#if defined(_OPENMP)
-		#pragma omp critical
-		#endif
-		{
-			for (unsigned i = 0; i < numComp; ++i) {
-				_localGamma[i] += localGamma_thread[i];
-			}
-		}
-	}
-
-	domainDecom->collCommInit(numComp);
-	for (unsigned i=0; i<numComp; i++){
-		domainDecom->collCommAppendDouble(_localGamma[i]);
-	}
-	domainDecom->collCommAllreduceSum();
-	for (unsigned i=0; i<numComp; i++){
-		_localGamma[i] = domainDecom->collCommGetDouble();
-	}
-	domainDecom->collCommFinalize();
-	for (unsigned i=0; i<numComp; i++){
-		_Gamma[i]+=_localGamma[i];
-	}
+void GammaWriter::calculateGamma(ParticleContainer* particleContainer, DomainDecompBase* domainDecom) {
+    std::vector<double> localGamma(_numComp, 0.0);
+
+    #if defined(_OPENMP)
+    #pragma omp parallel
+    #endif
+    {
+
+        std::vector<double> localGamma_thread(_numComp, 0.0);
+
+        for (auto tempMol = particleContainer->iterator(ParticleIterator::ONLY_INNER_AND_BOUNDARY); tempMol.isValid(); ++tempMol) {
+            // Only for particles within range
+            if ((tempMol->r(1) - _range.ymax)*(tempMol->r(1) - _range.ymin) <= 0) {
+                const unsigned cid = tempMol->componentid() + 1;
+                const double gamma = tempMol->Vi(1) - 0.5 * (tempMol->Vi(0) + tempMol->Vi(2));
+                localGamma_thread.at(cid) += gamma;
+                localGamma_thread.at(0)   += gamma;  // 0 is component-independent value
+            }
+        }
+
+        #if defined(_OPENMP)
+        #pragma omp critical
+        #endif
+        {
+            for (unsigned i = 0; i < _numComp; ++i) {
+                localGamma.at(i) += localGamma_thread.at(i);
+            }
+        }
```

Comment:
I tried to use the reduction. Please have a look as I am not very experienced with OpenMP. Furthermore, the particleContainer loop seems to be not supported by
`#pragma omp parallel for reduction(vec_double_plus: localGamma)`
therefore, I used
`#pragma omp parallel reduction(vec_double_plus: localGamma)`
I don't know if it is now faster/better compared to the old code...

Reply:
Yes, you need to use the pragma without `for`. Our iterators have to take care of the parallelism themselves because OpenMP can't deduce the loop count from them.